packages/amplify-cli-core/src/feature-flags/featureFlagProvider.ts

```diff
@@ -0,0 +1,422 @@
+      };
+
+      await JSONUtilities.writeJson(configFileName, config, {
+        keepComments: true,
+      });
+    }
+  };
+
+  private static ensureInitialized = (): void => {
+    if (!FeatureFlagProvider.instance) {
+      throw new Error('FeatureFlagsProvider is not initialized');
+    }
+  };
+
+  public static getBoolean = (flagName: string): boolean => {
+    return FeatureFlagProvider.getValue<boolean>(flagName, 'boolean');
+  };
+
+  public static getString = (flagName: string): string => {
+    return FeatureFlagProvider.getValue<string>(flagName, 'string');
+  };
+
+  public static getNumber = (flagName: string): number => {
+    return FeatureFlagProvider.getValue<number>(flagName, 'number');
+  };
+
+  public static getEffectiveFlags = (): Readonly<FeatureFlagsEntry> => {
+    FeatureFlagProvider.ensureInitialized();
+
+    return FeatureFlagProvider.instance.effectiveFlags;
+  };
+
+  public static getNewProjectDefaults = (): Readonly<FeatureFlagsEntry> => {
+    FeatureFlagProvider.ensureInitialized();
+
+    return FeatureFlagProvider.instance.newProjectDefaults;
+  };
+
+  public static getExistingProjectDefaults = (): Readonly<FeatureFlagsEntry> => {
+    FeatureFlagProvider.ensureInitialized();
+
+    return FeatureFlagProvider.instance.existingProjectDefaults;
+  };
+
+  private static getValue = <T>(flagName: string, type: FeatureFlagType): T => {
+    FeatureFlagProvider.ensureInitialized();
+
+    if (!flagName) {
+      throw new Error(`'flagName' argument is required`);
+    }
+
+    const parts = flagName.toLowerCase().split('.');
+
+    if (parts.length !== 2) {
+      throw new Error(`Invalid flagName value: '${flagName}'`);
+    }
+
+    // Get registration
+    const sectionRegistration = FeatureFlagProvider.instance.registrations.get(parts[0]);
+
+    if (!sectionRegistration) {
+      throw new Error(`Section '${parts[0]}' is not registered in feature provider`);
+    }
+
+    const flagRegistrationEntry = sectionRegistration.find(flag => flag.name === parts[1]);
+
+    if (!flagRegistrationEntry) {
+      throw new Error(`Flag '${parts[1]}' within '${parts[0]}' is not registered in feature provider`);
+    }
+
+    if (flagRegistrationEntry.type !== type) {
+      throw new Error(`'${flagName}' is a ${flagRegistrationEntry.type} type, not ${type}`);
+    }
+
+    let value: T | undefined;
+
+    // Check if effective values has a value for the requested flag
+    if (FeatureFlagProvider.instance.effectiveFlags[parts[0]] && FeatureFlagProvider.instance.effectiveFlags[parts[0]][parts[1]]) {
+      value = <T>FeatureFlagProvider.instance.effectiveFlags[parts[0]][parts[1]];
+    }
+
+    // If there is no value, return the registered defaults for existing projects
+    if (!value) {
+      value = <T>(flagRegistrationEntry.defaultValueForExistingProjects as unknown);
+    }
+
+    return value;
+  };
+
+  private buildJSONSchemaFromRegistrations = () => {
```

Comment:
this doesn't seem like something that should be in this class. Would recommend breaking out this logic into another file / class and depending on it.